### PR TITLE
Support redunce volume size

### PR DESCRIPTION
**Pull request by d-kuro**
https://github.com/cybozu-go/moco/issues/537

Support redunce volume size in MOCO.
It is implemented based on the proposals added below.

* https://github.com/cybozu-go/moco/pull/538

### CODE REVIEW

File: e2e/pvc_test.go [hunk 1]
```diff
@@ -24,6 +21,9 @@ var pvcTestYAML string
 var pvcApplyYAML string
 
 var _ = Context("pvc_test", func() {
+	if doUpgrade {
+		return
+	}
```
Comment (d-kuro) [inline e2e/pvc_test.go:26]: This scenario was not originally required to be run in the upgrade test, so we have changed it to skip the entire test.

As in the following test:
https://github.com/cybozu-go/moco/blob/b63186fc492f206813e1684abaecbbc0b4ff8c1e/e2e/lifecycle_test.go#L24

File: controllers/pvc.go [hunk 2]
```diff
@@ -174,7 +174,12 @@ func (*MySQLClusterReconciler) needResizePVC(cluster *mocov1beta2.MySQLCluster,
 		case i == 0: // volume size is equal
 			continue
 		case i == 1: // volume size is greater
-			return nil, false, fmt.Errorf("failed to resize pvc %q, want size: %s, deployed size: %s: %w", pvc.Name, wantSize, deployedSize, ErrReduceVolumeSize)
```
Comment (masa213f) [inline controllers/pvc.go:177]: Please update the following test in `controllers/pvc_test.go`.
https://github.com/cybozu-go/moco/blob/v0.16.1/controllers/pvc_test.go#L157-L163

Also, add a new test case. 
A MySQLCluster has 2 volume claim templates. One expands the volume size, and the other reduces it.

File: e2e/pvc_test.go [hunk 3]
```diff
@@ -66,107 +67,71 @@ var _ = Context("pvc_test", func() {
...
+	})
 
-			var pvcList corev1.PersistentVolumeClaimList
-			if err := json.Unmarshal(out, &pvcList); err != nil {
-				return err
-			}
-			if len(pvcList.Items) < 1 {
-				return errors.New("not found pvcs")
-			}
+	It("should volume size reduce succeed", func() {
```
Comment (masa213f) [inline e2e/pvc_test.go:102]: Can you confirm that the data is not lost after this test?

**Review by masa213f — CHANGES_REQUESTED** [review]

File: e2e/pvc_test.go (hunk 3 above)
Comment (d-kuro) [reply e2e/pvc_test.go:118]: Thanks, I added test to check for data loss.
https://github.com/cybozu-go/moco/pull/552/commits/57652b63c8237fa2a2fe4eaf2a8086a73507d0b1

File: controllers/pvc.go (hunk 2 above)
Comment (d-kuro) [reply controllers/pvc.go:177]: I added tests.
Also, there was a problem with the test code, which has been corrected.
https://github.com/cybozu-go/moco/pull/552/commits/32b8fbbd9cc1b4ea2b024e04f6f1a82a57282d2f

**Comment by masa213f** [comment]
LGTM. Thank you!

**Review by masa213f — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**